File: verilog/assembler/src/main/scala/AluOp.java
import java.util.Arrays;

public enum AluOp {
    // useful for zeroing RAM because we can't fetch a zero from ROM at same time
    ZERO(0),
    PASS_A(1),
    PASS_B(2),
    NEGATE_A(3),
    NEGATE_B(4),
    BA_DIV_10(5),
    BA_MOD_10(6),
    B_PLUS_1(7),  // REG = RAM + 1
    B_MINUS_1(8),  // REG = RAM - 1
    A_PLUS_B(9, "+"),
    A_MINUS_B(10, "-"),
    B_MINUS_A(11),
    A_MINUS_B_SIGNEDMAG(12),
    A_PLUS_B_PLUS_C(13),
    A_MINUS_B_MINUS_C(14),
    B_MINUS_A_MINUS_C(15),
    A_TIMES_B_LO(16, "*LO"),
    A_TIMES_B_HI(17, "*HI"),
    A_DIV_B(18, "/"),
    A_MOD_B(19, "%"),
    A_LSL_B(20, "<<"),
    A_LSR_B(21, ">>"),
    A_ASR_B(22, ">>>"),
    A_RLC_B(23, "<-"),
    A_RRC_B(24, "->"),
    A_AND_B(25, "&"),
    A_OR_B(26, "|"),
    A_XOR_B(27, "^"),
    A_NAND_B(28, "!&"),
    NOT_B(29),
    A_PLUS_B_BCD(20),
    A_MINUS_B_BCD(31);

    final int id;
    final String abbrev;

    AluOp(int id) {
        this.id = id;
        this.abbrev = null;
    }
    AluOp(int id, String abbrev) {
        this.id = id;
        this.abbrev = abbrev;
    }

    boolean isAbbreviated() {
        return abbrev != null;
    }
    static public AluOp valueOf(int id) {
        if (id > 31 || id < 0) throw new RuntimeException("unknown AluOp " + id);

        return Arrays.stream(values()).filter(c -> c.id == id).findFirst().get();
    }
    static public AluOp valueOfAbbrev(String abbrev) {
        return Arrays.stream(values()).filter(c -> c.abbrev.equals(abbrev)).findFirst().get();
    }
}

File: assembler/src/main/scala/asm/AluOp.scala
package asm

sealed class AluOp private(val id: Int, val abbrev: String = null) extends E {
    def isAbbreviated: Boolean = abbrev != null
}

object AluOp  {
    object ZERO extends AluOp(0)
    object PASS_A extends AluOp(1)
    object PASS_B extends AluOp(2)
    object NEGATE_A extends AluOp(3)
    object NEGATE_B extends AluOp(4)
    object BA_DIV_10 extends AluOp(5)
    object BA_MOD_10 extends AluOp(6)
    object B_PLUS_1 extends AluOp(7)  // REG = RAM + 1
    object B_MINUS_1 extends AluOp(8)  // REG = RAM - 1
    object A_PLUS_B extends AluOp(9, "+")
    object A_MINUS_B extends AluOp(10, "-")
    object B_MINUS_A extends AluOp(11)
    object A_MINUS_B_SIGNEDMAG extends AluOp(12)
    object A_PLUS_B_PLUS_C extends AluOp(13)
    object A_MINUS_B_MINUS_C extends AluOp(14)
    object B_MINUS_A_MINUS_C extends AluOp(15)
    object A_TIMES_B_LO extends AluOp(16, "*LO")
    object A_TIMES_B_HI extends AluOp(17, "*HI")
    object A_DIV_B extends AluOp(18, "/")
    object A_MOD_B extends AluOp(19, "%")
    object A_LSL_B extends AluOp(20, "<<")
    object A_LSR_B extends AluOp(21, ">>")
    object A_ASR_B extends AluOp(22, ">>>")
    object A_RLC_B extends AluOp(23, "<-")
    object A_RRC_B extends AluOp(24, "->")
    object A_AND_B extends AluOp(25, "&")
    object A_OR_B extends AluOp(26, "|")
    object A_XOR_B extends AluOp(27, "^")
    object A_NAND_B extends AluOp(28, "!&")
    object NOT_B extends AluOp(29)
    object A_PLUS_B_BCD extends AluOp(20)
    object A_MINUS_B_BCD extends AluOp(31)

    def valueOf(id: Int) : AluOp = {
        values.find(c => c.id == id).getOrElse(throw new RuntimeException("unknown AluOp " + id))
    }
    def  valueOfAbbrev(abbrev: String): AluOp = {
        values.find(c => c.abbrev.equals(abbrev)).getOrElse(throw new RuntimeException("unknown AluOp abbrev '" + abbrev + "'"))
    }

    def values: Seq[AluOp ] = Seq(
          ZERO ,
          PASS_A ,
          PASS_B ,
          NEGATE_A ,
          NEGATE_B ,
          BA_DIV_10 ,
          BA_MOD_10 ,
          B_PLUS_1 ,
          B_MINUS_1 ,
          A_PLUS_B ,
          A_MINUS_B ,
          B_MINUS_A ,
          A_MINUS_B_SIGNEDMAG ,
          A_PLUS_B_PLUS_C ,
          A_MINUS_B_MINUS_C ,
          B_MINUS_A_MINUS_C ,
          A_TIMES_B_LO ,
          A_TIMES_B_HI ,
          A_DIV_B ,
          A_MOD_B ,
          A_LSL_B ,
          A_LSR_B ,
          A_ASR_B ,
          A_RLC_B ,
          A_RRC_B ,
          A_AND_B ,
          A_OR_B ,
          A_XOR_B ,
          A_NAND_B ,
          NOT_B ,
          A_PLUS_B_BCD ,
          A_MINUS_B_BCD
      )
}
